### ISSUE TITLE
SMART_REJECT/TCP-RESET is considered harmful based on RFC3360

### ISSUE THREAD

**Issue by andvgal** [firehol/firehol#338]
Implementing netdata support with firehol **3.1.5**, I've noticed python plugins are stuck in `SYN-SENT` state with fine controlled `lo` interface.

Further debugging shows that the `SMART_REJECT` rule sends `TCP-RESET` which is getting dropped by input rule on the local interface. _Perhaps, you guys just accept all traffic on `lo` and do not see the problem. I've used to just ignore tons of related firewall logs previously._

Digging further, it seems TCP-RESET on firewall is strictly against current best practices defined in RFC3360: https://tools.ietf.org/html/rfc3360

IMO, there are the following actions to take:

1. Make the SMART_REJECT configurable **disabled-by-default** feature to be fully compliant with the RFC.
2. Ensure to accept all RST packets for both IN and OUT at least for "policy reject" interfaces - be compatible with non-compliant solutions without jeopardizing "policy drop" cases.
3. Update https://firehol.org/firehol-manual/firehol-actions/ to clearly reference the RFC

**Your thoughts?**

P.S. Just for reference, my current solution is to add the following at the end of the `interface` in question:
```
    # Note: icmp-port-unreachable is not recognized by IPv6 iptabes
    policy reject with port-unreach
    # ....
    # prevent noise & local connection timeouts
    iptables -A in_local -p tcp --tcp-flags RST RST -j ACCEPT
    iptables -A out_local -p tcp --tcp-flags RST RST -j ACCEPT
    ip6tables -A in_local -p tcp --tcp-flags RST RST -j ACCEPT
    ip6tables -A out_local -p tcp --tcp-flags RST RST -j ACCEPT
```

P.P.S. I've seen not-yet-released `FIREHOL_ACCEPT_OUTPUT_UNMATCHED_TCP_RST`, but it's not sufficient.

**Comment by andvgal** [comment]
@philwhineray any comments?

**Comment by philwhineray** [comment]
Hi, sorry for the lack of reply until now.

TLDR; I would certainly accept a documentation patch, and possibly a code
patch. If you don't have the time / inclination, we can leave this open as
a "when I get around to it" which has the benefit of being searchable by
anyone else suffering similarly: thanks for your comprehensive report.

I don't know anyone else running a firewall on `lo`, so you're correct,
this is kind of a special case that probably few are hitting (or they are
working around).

I read RFC3360 quickly (so I may have missed some nuance) but as far as I
can tell, it is about middle-boxes using the reset as a general defense. I
think FireHOL is fine in situations where the host is not both input and
output, since it uses this only to indicate a server port is closed when a
client tries to establish an unwanted connection, nothing more.

WIth `lo`, we are both client and server, and so now we have a problem as I
understand it, because we have prevented the outgoing connection but then
dropped the packet we generates which was to  to inform the client that its
connection has been rejected.

I wouldn't recommend allowing RST packets through regardless, that is the
kind of weakening that might lead to an adversary probing past a firewall
which would otherwise be hiding hosts. I can certainly see a case for
allowing it on loopback interfaces though, certainly as an option, and
possibly by default.

Cheers
Phil

**Comment by andvgal** [comment]
I can spent some time provided a PR, if we settle technical requirements for the outcome.

I've seen similar issues with delays in infrastructure when RSTs are sent by OS due to other cases. Perhaps, it can be some type of command ( e.g. "'accept-tcp-reset").

**Comment by andvgal** [comment]
Slow-down is caused on client due to smart-reject using TCP-RST. That leads to **client-side** to filter out the TCP-RST which does not belong to any established connection by fact. As a consequence, connection just timeouts instead of getting "port unreachable" or similar ICMP error. With short timeouts of 3-5 seconds it was not so painful, but netdata revealed the problem.

Actually, the fix would be to disable smart-reject first of all. However, this is not controllable yet. That's step 1 of the description.

The TCP-RST part is another topic to be compatible with RFC3360 violating third-party nodes. That's step 2 of the description.

**Comment by philwhineray** [comment]
I still don't think the RFC applies here - it is taking about sending RST in response to unknown conditions which prevents extensions to the TCP protocol. Smart reset sends RST in the very specific case the firewall wants to prevent the connection being established, which is equivalent to the destination reporting that the port does not have anything listening on it. Paragraph 4 of section 6 (the firewalled-off TCP port sentence):

~~~~
We would note that this document does not consider middleboxes that
block complete transport protocols.  We also note that this document
is not addressing firewalls that send resets in response to a TCP SYN
packet to a firewalled-off TCP port.  Such a use of resets seems
consistent with the semantics of TCP reset.  This document is only
considering the problems caused by middleboxes that block specific
packets within a transport protocol when other packets from that
transport protocol are forwarded by the middlebox unaltered.
~~~~

You could disable it easily enough - and to test it you just need to to an iptables-save, edit the output and reload - but I don't think that will help the client. If the SYN is dropped with no response, then the client will retransmit and then timeout anyway, won't it? So are you proposing to replace it somehow?

Allowing the RST through, or blocking the outgoing SYN in the INPUT chain seem like the only ways you could avoid the timeout to me. But to allow the RST you need something to send it in the first place, so what will do that if not smart reject? It would be sent by the networking code if the connection were permitted by the firewall and the port was closed but the way to achieve that is with server accept rules in the firewall.

Or maybe I have missed something would be helpful if you could provide an easily reproduced test script (maybe netcat listening and connecting) with rules as produced leading to the timeout and rules as you need them to be to avoid it... but I think in outline that's what you were describing in your original post.

**Comment by philwhineray** [comment]
I see that you mentioned ICMP unreachable - but my understanding is that the behaviour of TCP in response to that is itself not very well defined, which is the reason the RST is being used in the first place.

I think you could do this to achieve the effect (i.e. smart code will never get triggered):

~~~~
  server4 all reject with icmp-port-unreachable
  server6 all reject with icmp6-port-unreachable
~~~~

If it works for you, the best patch would be to allow providing ipv6 and ipv4 policy separately. You can already specify the rejection type but it looks like the code gets confused and tries to send the same icmp type to both versions of iptables, which does not work.
